website/addons/owncloud/model.py

```diff
@@ -0,0 +1,148 @@
+# -*- coding: utf-8 -*-
+from urllib2 import HTTPError
+import logging
+from modularodm import fields
+from framework.auth import Auth
+from website.addons.base import exceptions
+from website.addons.base import StorageAddonBase
+from website.addons.base import (
+    AddonOAuthNodeSettingsBase, AddonOAuthUserSettingsBase,
+)
+from website.addons.owncloud.serializer import OwnCloudSerializer
+from website.addons.owncloud.utils import (
+    ExternalAccountConverter, OwnCloudNodeLogger
+)
+
+logger = logging.getLogger(__name__)
+
+class OwnCloudProvider(object):
+    """An alternative to `ExternalProvider` not tied to OAuth"""
+
+    name = 'ownCloud'
+    short_name = 'owncloud'
+    serializer = OwnCloudSerializer
+
+    def __init__(self, account=None):
+        super(OwnCloudProvider, self).__init__()
+        self.account = account
+
+    def __repr__(self):
+        return '<{name}: {status}>'.format(
+            name=self.__class__.__name__,
+            status=self.account.display_name if self.account else 'anonymous'
+        )
+
+class AddonOwnCloudUserSettings(AddonOAuthUserSettingsBase):
+    oauth_provider = OwnCloudProvider
+    serializer = OwnCloudSerializer
+
+class AddonOwnCloudNodeSettings(StorageAddonBase, AddonOAuthNodeSettingsBase):
+    oauth_provider = OwnCloudProvider
+    serializer = OwnCloudSerializer
+
+    folder_name = fields.StringField()
+
+    @property
+    def complete(self):
+        return bool(self.has_auth and self.folder_name is not None)
+
+    @property
+    def folder_id(self):
+        return self.folder_name
+
+    @property
+    def folder_path(self):
+        return self.folder_name
+
+    @property
+    def nodelogger(self):
+        auth = None
+        if self.user_settings:
+            auth = Auth(self.user_settings.owner)
+        return OwnCloudNodeLogger(
+            node=self.owner,
+            auth=auth
+        )
+
+    def set_folder(self, folder, auth=None):
+        if folder == '/ (Full ownCloud)':
```

Comment:
This class should have a `def fetch_folder_name` that reverses this conversion, so that the front-end doesn't display `/` as the linked folder (and for `generic_view.root_folder` support).